### PR TITLE
Move createDatabase to RP

### PR DESCRIPTION
**Pull request by victor-meng**
- Refactored and moved all code paths for creating database in different APIs into createDatabase.ts.
- Before creating a database with RP, we need to check whether there already exists a database with the same id by using the getDatabase call. This is because RP bundles the create call with the update call which means if we try to create a database that already exists, RP will update the existing database instead of creating a new one.
- If there's no existing database with the same id, then RP returns 404 for the get call, so we need to add a try catch block to intercept the error and check if the error code is `"NotFound"`

### CODE REVIEW

File: src/Common/dataAccess/createDatabase.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    params.databaseId,
+    rpPayload
+  );
+  return createResponse && (createResponse.properties.resource as DataModels.Database);
+}
+
+async function createDatabaseWithSDK(params: DataModels.CreateDatabaseParams): Promise<DataModels.Database> {
+  const createBody: DatabaseRequest = { id: params.databaseId };
+  const databaseOptions: RequestOptions = {};
+  // TODO: replace when SDK support autopilot
```
Comment (southpolesteve) [inline src/Common/dataAccess/createDatabase.ts:222]: This should work now. maxThroughput is available first class in the SDK. Ping @zfoster if you hit any trouble

File: src/Common/dataAccess/createDatabase.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+}
+
+function constructRpOptions(params: DataModels.CreateDatabaseParams): CreateUpdateOptions {
+  if (!params.databaseLevelThroughput) {
+    return {};
+  }
+
+  if (params.autoPilotMaxThroughput) {
+    return {
+      [Constants.HttpHeaders.autoPilotThroughput]: { maxThroughput: params.autoPilotMaxThroughput }
```
Comment (southpolesteve) [inline src/Common/dataAccess/createDatabase.ts:244]: Same as above, I think you don't need to set this header anymore.

File: src/Common/dataAccess/createDatabase.ts [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  createUpdateGremlinDatabase,
+  getGremlinDatabase
+} from "../../Utils/arm/generatedClients/2020-04-01/gremlinResources";
+import { logConsoleProgress, logConsoleError, logConsoleInfo } from "../../Utils/NotificationConsoleUtils";
+import { logError } from "../Logger";
+import { refreshCachedOffers, refreshCachedResources } from "../DataAccessUtilityBase";
+import { sendNotificationForError } from "./sendNotificationForError";
+import { userContext } from "../../UserContext";
+
+export async function createDatabase(params: DataModels.CreateDatabaseParams): Promise<DataModels.Database> {
```
Comment (languy) [inline src/Common/dataAccess/createDatabase.ts:32]: Can we switch from the old-school `function` syntax to the fat arrow syntax:
`export const createDatabase = async () => { }`;

**Review by languy — APPROVED** [review]

File: src/Common/dataAccess/createDatabase.ts (hunk 2 above)
Comment (victor-meng) [reply src/Common/dataAccess/createDatabase.ts]: This is actually a bug. I'm setting the RP options for creating database so I shouldn't be using the header string. Fixed in my latest commit.

**Comment by victor-meng** [comment]
> After this change is `AddDatabaseUtility` used anymore? If not, should delete it.

@southpolesteve Some of the functions in there are being used but we should be able to clean them up. I can do this in another PR. I think we should be able to remove `AddCollectionUtility` too after we move create collection to RP. I'll probably clean up both utilities in the same PR.

File: src/Common/dataAccess/createDatabase.ts [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  getMongoDBDatabase
+} from "../../Utils/arm/generatedClients/2020-04-01/mongoDBResources";
+import {
+  createUpdateGremlinDatabase,
+  getGremlinDatabase
+} from "../../Utils/arm/generatedClients/2020-04-01/gremlinResources";
+import { logConsoleProgress, logConsoleError, logConsoleInfo } from "../../Utils/NotificationConsoleUtils";
+import { logError } from "../Logger";
+import { refreshCachedOffers, refreshCachedResources } from "../DataAccessUtilityBase";
+import { sendNotificationForError } from "./sendNotificationForError";
```
[the remark this answers is not in the document]
Comment (victor-meng) [reply src/Common/dataAccess/createDatabase.ts:28]: I will update all of them in a separate PR. Steve suggested that we could add a lint rule for this.

File: src/Common/dataAccess/createDatabase.ts (hunk 3 above)
Comment (victor-meng) [reply src/Common/dataAccess/createDatabase.ts:31]: I'll make the switch in the same PR where I update the module names.

File: src/Common/dataAccess/createDatabase.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+      throw new Error(`Create database failed: database with id ${params.databaseId} already exists`);
+    }
+  } catch (error) {
+    if (error.code !== "NotFound") {
+      throw error;
+    }
+  }
+
+  const options: CreateUpdateOptions = constructRpOptions(params);
+  const rpPayload: SqlDatabaseCreateUpdateParameters = {
```
Comment (ato9000) [inline src/Common/dataAccess/createDatabase.ts:92]: nit: if all APIs use the same parameters, consider renaming this interface to DatabaseCreateUpdateParameters

Comment (victor-meng) [reply src/Common/dataAccess/createDatabase.ts:92]: This is actually a bug. Different APIs should have different parameter types. I'll update this in my next PR. Thanks for catching it!

**Pull request merged**